tests/kernel/protection/src/main.c

```diff
@@ -0,0 +1,181 @@
+{
+	return (i + 1);
+}
+
+static void execute_from_buffer(u8_t *dst)
+{
+	void *src = FUNC_TO_PTR(add_one);
+	int (*func)(int i) = PTR_TO_FUNC(dst);
+	int i = 1;
+
+	/* Copy add_one() code to destination buffer. */
+	memcpy(dst, src, BUF_SIZE);
+	DO_BARRIERS();
+
+	/*
+	 * Try executing from buffer we just filled.
+	 * Optimally, this triggers a fault.
+	 * If not, we check to see if the function
+	 * returned the expected result as confirmation
+	 * that we truly executed the code we wrote.
+	 */
+	INFO("trying to call code written to %p\n", func);
+	i = func(i);
+	INFO("returned from code at %p\n", func);
+	if (i == 2) {
+		INFO("Execute from target buffer succeeded!\n");
+	} else {
+		INFO("Did not get expected return value!\n");
+	}
+}
+
+static void write_ro(void *unused1, void *unused2, void *unused3)
+{
+	u32_t *ptr = (u32_t *)&rodata_var;
+
+	/*
+	 * Try writing to rodata.  Optimally, this triggers a fault.
+	 * If not, we check to see if the rodata value actually changed.
+	 */
+	INFO("trying to write to rodata at %p\n", ptr);
+	*ptr = ~RODATA_VALUE;
+
+	DO_BARRIERS();
+
+	if (*ptr == RODATA_VALUE) {
+		INFO("rodata value still the same\n");
+	} else if (*ptr == ~RODATA_VALUE) {
+		INFO("rodata modified!\n");
+	} else {
+		INFO("something went wrong!\n");
+	}
+
+	zassert_unreachable("Write to rodata did not fault");
+}
+
+static void write_text(void *unused1, void *unused2, void *unused3)
+{
+	void *src = FUNC_TO_PTR(add_one);
+	void *dst = FUNC_TO_PTR(overwrite_target);
+	int i = 1;
+
+	/*
+	 * Try writing to a function in the text section.
+	 * Optimally, this triggers a fault.
+	 * If not, we try calling the function after overwriting
+	 * to see if it returns the expected result as
+	 * confirmation that we truly executed the code we wrote.
+	 */
+	INFO("trying to write to text at %p\n", dst);
+	memcpy(dst, src, BUF_SIZE);
+	DO_BARRIERS();
+	i = overwrite_target(i);
+	if (i == 2) {
+		INFO("Overwrite of text succeeded!\n");
+	} else {
+		INFO("Did not get expected return value!\n");
+	}
+
+	zassert_unreachable("Write to text did not fault");
+}
+
+static void exec_data(void *unused1, void *unused2, void *unused3)
+{
+	execute_from_buffer(data_buf);
+	zassert_unreachable("Execute from data did not fault");
+}
+
+static void exec_stack(void *unused1, void *unused2, void *unused3)
+{
+	u8_t stack_buf[BUF_SIZE] __aligned(sizeof(int));
+
+	execute_from_buffer(stack_buf);
+	zassert_unreachable("Execute from stack did not fault");
+}
+
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+static void exec_heap(void *unused1, void *unused2, void *unused3)
+{
+	u8_t *heap_buf = k_malloc(BUF_SIZE);
+
+	execute_from_buffer(heap_buf);
+	k_free(heap_buf);
+	zassert_unreachable("Execute from heap did not fault");
+}
+#endif
+
+static void do_run_test(const char *name, void (*f)(void *, void *, void *))
+{
+	k_thread_create(&test_thread, thread_stack, STACKSIZE,
+			f, NULL, NULL, NULL,
+			K_PRIO_COOP(PRIORITY), 0,
+			K_NO_WAIT);
+}
```

Comment:
Since ztest creates threads internally for each test, we could avoid the need to create this separate thread if ztest provided a ztest_test_pass() analogous to its ztest_test_fail() function, i.e. one that sets test_result = 0 and calls k_sem_give(&test_end_signal). Then we could call that from _SysFatalErrorHandler() upon the MPU fault.  Otherwise, we end up hanging on the k_sem_take(&test_end_signal, K_FOREVER) cal in run_test().  If desired, I could add that and include it as part of this commit.